test/performance/thread_safe_jit.cpp

```diff
@@ -0,0 +1,125 @@
+#include "Halide.h"
+#include <cstdio>
+#include <functional>
+#include <thread>
+#include "halide_benchmark.h"
+
+/** \file Test to demonstrate using JIT across multiple threads with
+ * varying parameters passed to realizations. Performance is tested
+ * by comparing a technique that recompiles vs one that should not.
+ */
+
+using namespace Halide;
+using namespace Halide::Tools;
+
+struct test_func {
+    Param<int32_t> p;
+    ImageParam in{Int(32), 1};
+    Func f;
+    Var x;
+
+    test_func() {
+        Expr big = 0;
+        for (int i = 0; i < 75; i++) {
+            big += p;
+        }
+        Func inner;
+        inner(x) = x * in(clamp(x, 0 , 9)) + big;
+        f(x) = inner(x - 1) + inner(x) + inner(x + 1);
+        inner.compute_at(f, x);
+    }
+};
+
+// The Halide compiler is currently not guaranteed to be thread safe.
+std::mutex compiler_mutex;
+
+Buffer<int32_t> bufs[16];
+
+void separate_func_per_thread_executor(int index) {
+    test_func test;
+
+    {
+        std::lock_guard<std::mutex> lock(compiler_mutex);
+
+        test.f.compile_jit();
+    }
+
+    test.p.set(index);
+    test.in.set(bufs[index]);
+    for (int i = 0; i < 10; i++) {
+       Buffer<int32_t> result = test.f.realize(10);
+        for (int j = 0; j < 10; j++) {
+            int32_t left = (j - 1) * bufs[index](std::min(std::max(0, j - 1), 9)) + 75 * index;
+            int32_t middle = j * bufs[index](std::min(std::max(0, j), 9)) + 75 * index;
+            int32_t right = (j + 1) * bufs[index](std::min(std::max(0, j + 1), 9)) + 75 * index;
+            assert(result(j) == left + middle + right);
+        }
+    }
+}
+
+void separate_func_per_thread() {
+    std::thread threads[16];
+
+    for (auto &thread : threads) {
+        thread = std::thread(separate_func_per_thread_executor,
+                             (int)(&thread - threads));
+    }
+
+    for (auto &thread : threads) {
+        thread.join();
+    }
+}
+
+void same_func_per_thread_executor(int index, test_func &test) {
+    for (int i = 0; i < 10; i++) {
+        Buffer<int32_t> result = test.f.realize(10, get_target_from_environment(),
+                                                { { test.p, index },
+                                                  { test.in, bufs[index] } });
+        for (int j = 0; j < 10; j++) {
+            int32_t left = (j - 1) * bufs[index](std::min(std::max(0, j - 1), 9)) + 75 * index;
+            int32_t middle = j * bufs[index](std::min(std::max(0, j), 9)) + 75 * index;
+            int32_t right = (j + 1) * bufs[index](std::min(std::max(0, j + 1), 9)) + 75 * index;
+            assert(result(j) == left + middle + right);
+        }
+    }
+}
+
+void same_func_per_thread() {
+    std::thread threads[16];
+    test_func test;
+
+    {
+        std::lock_guard<std::mutex> lock(compiler_mutex);
```

Comment:
This is unnecessary in this case I think. Harmless, but slightly misleading too.

Reply:
Added a comment to indicate why it is included even though in some sense it is not strictly necessary.